akka-actor-typed/src/main/scala/akka/actor/typed/internal/delivery/ProducerController.scala

```diff
@@ -0,0 +1,506 @@
+ * from the consumer that it has received and processed all messages up to that sequence number.
+ *
+ * The `ConsumerController` will send [[ProducerController.Internal.Resend]] if a lost message is detected
+ * and then the `ProducerController` will resend all messages from that sequence number. The producer keeps
+ * unconfirmed messages in a buffer to be able to resend them. The buffer size is limited
+ * by the request window size.
+ *
+ * The resending is optional, and the `ConsumerController` can be started with `resendLost=false`
+ * to ignore lost messages, and then the `ProducerController` will not buffer unconfirmed messages.
+ * In that mode it provides only flow control but no reliable delivery.
+ */
+object ProducerController {
+
+  sealed trait InternalCommand
+
+  sealed trait Command[A] extends InternalCommand
+
+  final case class Start[A](producer: ActorRef[RequestNext[A]]) extends Command[A]
+
+  final case class RequestNext[A](
+      producerId: String,
+      currentSeqNr: Long,
+      confirmedSeqNr: Long,
+      sendNextTo: ActorRef[A],
+      askNextTo: ActorRef[MessageWithConfirmation[A]])
+
+  final case class RegisterConsumer[A](consumerController: ActorRef[ConsumerController.Command[A]]) extends Command[A]
+
+  /**
+   * For sending confirmation message back to the producer when the message has been fully delivered, processed,
+   * and confirmed by the consumer. Typically used with `ask` from the producer.
+   */
+  final case class MessageWithConfirmation[A](message: A, replyTo: ActorRef[Long]) extends InternalCommand
+
+  object Internal {
+    final case class Request(confirmedSeqNr: Long, upToSeqNr: Long, supportResend: Boolean, viaTimeout: Boolean)
+        extends InternalCommand {
+      require(confirmedSeqNr <= upToSeqNr, s"confirmedSeqNr [$confirmedSeqNr] should be <= upToSeqNr [$upToSeqNr]")
+    }
+    final case class Resend(fromSeqNr: Long) extends InternalCommand
+    final case class Ack(confirmedSeqNr: Long) extends InternalCommand
+  }
+
+  private case class Msg[A](msg: A) extends InternalCommand
+  private case object ResendFirst extends InternalCommand
+
+  private case class LoadStateReply[A](state: DurableProducerQueue.State[A]) extends InternalCommand
+  private case class LoadStateFailed(attempt: Int) extends InternalCommand
+  private case class StoreMessageSentReply(ack: DurableProducerQueue.StoreMessageSentAck)
+  private case class StoreMessageSentFailed[A](messageSent: DurableProducerQueue.MessageSent[A], attempt: Int)
+      extends InternalCommand
+
+  private case class StoreMessageSentCompleted[A](messageSent: DurableProducerQueue.MessageSent[A])
+      extends InternalCommand
+
+  private final case class State[A](
+      requested: Boolean,
+      currentSeqNr: Long,
+      confirmedSeqNr: Long,
+      requestedSeqNr: Long,
+      pendingReplies: Vector[(Long, ActorRef[Long])],
+      supportResend: Boolean,
```

Comment:
This confuses me a bit, that it is both in the state of the producer, and in the Request message, whether to support resends. For example on receive `Request` half of the logic (in `onAck`) uses the flag from state, and half from the message. Would it ever change after the PC has been started?

Reply:
I wanted to drive the setting from the CC. It would be more intutive if it was passed to PC in the RegisterConsumer instead of in each Request, but if I remember correctly that doesn't work for the sharding scenario, because then there is no RegisterConsumer. Then they are connected via the fist SequencedMessage since PC sends via sharding.

You are right that it doesn't change. I'll look at the logic if something can be cleaned up.